File: src/main/webapp/js/projects.js
var projects = {};
var users = [];
var inProgressTasks = [];
var newTasks = [];
var doneTasks = [];
var currentProject = null;

initializer.successCallback = loadProjects;
$(document).ready(selectButtonOnLoad("#nav-projects"));

function fillUsersTable() {
    var table = $("#project-users");

    table.append('<tr>' +
        '<th width="250">Name</th>' +
        '<th width="50">Action</th>' +
        '</tr>');

    for (var i = 0; i < users.length; i++) {
        var user = users[i];

        table.append('<tr class=' + (i % 2 === 0 ? '"light-row"' : '"dark-row"') + '>' +
            '<td>' + user.firstName + ' ' + user.lastName + '</td>' +
            '<td>' +
            '<button class="button red-button small-button" title="Delete task" ' +
            'onclick="">' +
            '<img height="9" src="img/delete.svg"/>' +
            '</button>' +
            '</td>' +
            '</tr>');
    }
}

function loadUsers(projectId) {
    var projectUsers = $("#project-users");
    var usersText = $("#users-text");

    projectUsers.empty();

    // TODO load from backend
    if (projectId === "1") {
        users = [{
            id: 11,
            username: "user1",
            firstName: "User",
            lastName: "Test",
            role: "ROLE_ADMIN"
        }, {
            id: 22,
            username: "user2",
            firstName: "UUser",
            lastName: "TTest",
            role: "ROLE_USER"
        }];
    } else {
        users = [];
    }

    if (users.length === 0) {
        projectUsers.hide();
        usersText.show();
    } else {
        fillUsersTable();
        usersText.hide();
    }

    projectUsers.show();
}

function clearTasks() {
    $("#in-progress-tasks").empty();
    $("#new-tasks").empty();
    $("#done-tasks").empty();
}

function showNoTasks() {
    $("#in-progress-tasks").hide();
    $("#new-tasks").hide();
    $("#done-tasks").hide();
    $("#in-progress-label").hide();
    $("#new-label").hide();
    $("#done-label").hide();
    $("#in-progress-text").hide();
    $("#new-text").hide();
    $("#done-text").hide();
    $("#no-tasks-label").show();
}

function showTasks() {
    $("#in-progress-tasks").show();
    $("#new-tasks").show();
    $("#done-tasks").show();
    $("#in-progress-label").show();
    $("#new-label").show();
    $("#done-label").show();
    $("#no-tasks-label").hide();
}

function fillTaskTable(tableIdentifier, tasks, util) {
    var table = $(tableIdentifier);

    table.append('<tr>' +
        '<th width="80pt">ID</th>' +
        '<th width="80pt">Priority</th>' +
        '<th>Subject</th>' +
        '<th width="140pt">Assignee</th>' +
        '<th width="140pt">' + util.firstField + '</th>' +
        '<th width="140pt">' + util.secondField + '</th>' +
        '<th width="140pt">Action</th>' +
        '</tr>');

    for (var i = 0; i < tasks.length; i++) {
        var task = tasks[i];

        table.append('<tr class=' + (i % 2 === 0 ? '"light-row"' : '"dark-row"') + '>' +
            '<td>#' + task.id + '</td>' +
            '<td>' +
            '<div class="priority-div">' +
            prioritySpans(task.priority, MAX_PRIORITY) +
            '</div>' +
            '</td>' +
            '<td>' + task.subject + '</td>' +
            '<td>' + toString(task.assignee) + '</td>' +
            util.filler(task, i) +
            '</tr>');
    }
}

function loadTasks(projectId) {
    clearTasks();

    // TODO load from backend
    if (projectId === "1") {
        inProgressTasks = [{
            id: 555,
            priority: 3,
            subject: "test task 1",
            assignee: "user 1",
            createdTimestamp: 1487710619000,
            startedTimestamp: 1487710619000,
            dueTimestamp: 1487710619000,
            finishedTimestamp: 1487710619000,
            status: "IN_PROGRESS"
        }, {
            id: 2342,
            priority: 2,
            subject: "test task 2",
            assignee: null,
            createdTimestamp: 1487710619000,
            startedTimestamp: 1487710619000,
            dueTimestamp: 1487710619000,
            finishedTimestamp: 1487710619000,
            status: "IN_PROGRESS"
        }];

        newTasks = [{
            id: 235,
            priority: 4,
            subject: "new task 1",
            assignee: "user 2",
            createdTimestamp: 1487710619000,
            startedTimestamp: 1487710619000,
            dueTimestamp: 1487710619000,
            finishedTimestamp: 1487710619000,
            status: "NEW"
        }];

        doneTasks = [{
            id: 999,
            priority: 1,
            subject: "done task",
            assignee: "user 2",
            createdTimestamp: 1487710619000,
            startedTimestamp: 1487710619000,
            dueTimestamp: 1487710619000,
            finishedTimestamp: 1487710619000,
            status: "DONE"
        }];
    } else {
        inProgressTasks = [];
        newTasks = [];
        doneTasks = [];
    }

    if (inProgressTasks.length === 0 && newTasks.length === 0 && doneTasks.length === 0) {
        showNoTasks();
    } else {
        if (inProgressTasks.length === 0) {
            $("#in-progress-text").show();
        } else {
            $("#in-progress-text").hide();
            fillTaskTable("#in-progress-tasks", inProgressTasks, inProgressTasksUtil);
        }

        if (newTasks.length === 0) {
            $("#new-text").show();
        } else {
            $("#new-text").hide();
            fillTaskTable("#new-tasks", newTasks, newTasksUtil);
        }

        if (doneTasks.length === 0) {
            $("#done-text").show();
        } else {
            $("#done-text").hide();
            fillTaskTable("#done-tasks", doneTasks, doneTasksUtil);
        }

        showTasks();
    }
}

function loadProjects() {
    projects = {
        "1": {
            id: 1,
            name: "project 1",
            description: "this is a test project with tasks",
            owner: "user 1",
            projectStatus: "ACTIVE"
        }, "2": {
            id: 2,
            name: "project 2",
            description: "this is a test project without tasks",
            owner: "user 1",
            projectStatus: "ACTIVE"
        }
    }; // TODO: load from backend

    var projectTiles = $("#project-tiles");

    projectTiles.empty();

    for (var projectKey in projects) {
        if (projects.hasOwnProperty(projectKey)) {
            var project = projects[projectKey];

            projectTiles.append('<div id="project-' + project.id + '" onclick="selectProject(\'' + project.id + '\');"'
                + ' class="project-tile"><p>' + project.name + '</p></div>');
        }
    }
}

function prioritySpans(value, max) {
    var output = "";

    for (var i = 1; i <= max; i++) {
        if (i <= value) {
            output += '<span class="activated-span"></span>';
        } else {
            output += '<span></span>';
        }
    }

    return output;
}

function toString(string) {
    if (string === null) {
        return "-";
    } else {
        return string;
    }
}

function toDate(string) {
    if (string === null) {
        return "-";
    } else {
        var date = new Date(string);
        return date.getHours() + ":" + date.getMinutes() + " "
            + date.getDate() + "-" + (date.getMonth() + 1) + "-" + date.getFullYear();
    }
}

var inProgressTasksUtil = {
    firstField: "Started",
    secondField: "Due",
    filler: function(task, taskIndex) {
        return '<td>' + toDate(task.startedTimestamp) + '</td>' +
            '<td>' + toDate(task.dueTimestamp) + '</td>' +
            '<td>' +
            '<button class="button green-button small-button" title="Mark as done">' +
            '<img height="12" src="img/done.svg"/>' +
            '</button> ' +
            '<button class="button blue-button small-button" ' +
            'title="Assign task" ' +
            'onclick="showAssignModalInProgress(' + taskIndex + ');">' +
            '<img height="12" src="img/take.svg"/>' +
            '</button> ' +
            '<button class="button blue-button small-button" title="Edit task" ' +
            'onclick="showEditModalInProgress(' + taskIndex + ');">' +
            '<img height="12" src="img/edit.svg"/>' +
            '</button> ' +
            '<button class="button red-button small-button" title="Delete task" ' +
            'onclick="showDeleteModalInProgress(' + taskIndex + ');">' +
            '<img height="12" src="img/delete.svg"/>' +
            '</button>' +
            '</td>';
    }
};

var newTasksUtil = {
    firstField: "Created",
    secondField: "Due",
    filler: function(task, taskIndex) {
        return '<td>' + toDate(task.createdTimestamp) + '</td>' +
            '<td>' + toDate(task.dueTimestamp) + '</td>' +
            '<td>' +
            '<button class="button green-button small-button" title="Start task">' +
            '<img height="12" src="img/start.svg"/>' +
            '</button> ' +
            '<button class="button blue-button small-button" title="Assign task" ' +
            'onclick="showAssignModalNew(' + taskIndex + ');">' +
            '<img height="12" src="img/take.svg"/>' +
            '</button> ' +
            '<button class="button blue-button small-button" title="Edit task" ' +
            'onclick="showEditModalNew(' + taskIndex + ');">' +
            '<img height="12" src="img/edit.svg"/>' +
            '</button> ' +
            '<button class="button red-button small-button" title="Delete task" ' +
            'onclick="showDeleteModalNew(' + taskIndex + ');">' +
            '<img height="12" src="img/delete.svg"/>' +
            '</button>' +
            '</td>';
    }
};

var doneTasksUtil = {
    firstField: "Started",
    secondField: "Finished",
    filler: function(task, taskIndex) {
        return '<td>' + toDate(task.startedTimestamp) + '</td>' +
            '<td>' + toDate(task.finishedTimestamp) + '</td>' +
            '<td>' +
            '<button class="button blue-button small-button" title="Restart task">' +
            '<img height="12" src="img/restart.svg"/>' +
            '</button> ' +
            '<button class="button red-button small-button" title="Delete task" ' +
            'onclick="showDeleteModalDone(' + taskIndex + ');">' +
            '<img height="12" src="img/delete.svg"/>' +
            '</button>' +
            '</td>';
    }
};

function showAssignModal(task) {
    var dataList = $("#assignable-users");

    dataList.empty();

    for (var i = 0; i < users.length; i++) {
        var user = users[i];

        dataList.append('<option value="' + user.firstName + ' ' + user.lastName + '">' + user.username + '</option>');
    }

    $("#assign-task-name").text(task.subject);
    $("#assign-task-modal").show();
}

function showAssignModalInProgress(taskIndex) {
    showAssignModal(inProgressTasks[taskIndex]);
}

function showAssignModalNew(taskIndex) {
    showAssignModal(newTasks[taskIndex]);
}

function showEditModal(task) {
    $("#edit-task-modal").show();
    $("#task-subject").val(task.subject);
    $("#task-due").val(task.dueTimestamp);
    $("#task-priority").val(task.priority);
}

function showEditModalInProgress(taskIndex) {
    showEditModal(inProgressTasks[taskIndex]);
}

function showEditModalNew(taskIndex) {
    showEditModal(newTasks[taskIndex]);
}

function showDeleteModal(task) {
    $("#delete-task-modal").show();
    $("#delete-task-name").text(task.subject);
}

function showDeleteModalInProgress(taskIndex) {
    showDeleteModal(inProgressTasks[taskIndex]);
}

function showDeleteModalNew(taskIndex) {
    showDeleteModal(newTasks[taskIndex]);
}

function showDeleteModalDone(taskIndex) {
    showDeleteModal(doneTasks[taskIndex]);
}

function selectProject(projectId) {
    if (currentProject === projectId) {
        return;
    }

    $("#project-" + currentProject).removeClass("selected-project-tile");

    currentProject = projectId;

    loadTasks(projectId);
    loadUsers(projectId);

    $("#project-" + projectId).addClass("selected-project-tile");
    $("#project-name").text(projects[projectId].name);
    $("#project-description").text(projects[projectId].description);
    $("#no-project").hide();
    $("#selected-project").show();
}

function selectTab(tabId, otherId) {
    var tab = $(tabId);
    var otherTab = $(otherId);

    if (!tab.hasClass("selected-project-tab")) {
        tab.removeClass("project-tab");
        tab.addClass("selected-project-tab");
        otherTab.removeClass("selected-project-tab");
        otherTab.addClass("project-tab");
    }
}

function showOverviewTab() {
    selectTab("#overview-tab", "#tasks-tab");

    $("#overview-panel").show();
    $("#tasks-panel").hide();
}

function showTasksTab() {
    selectTab("#tasks-tab", "#overview-tab");

    $("#tasks-panel").show();
    $("#overview-panel").hide();
}

File: src/main/scala/at/doml/taskmanager/pages/Projects.scala
package at.doml.taskmanager.pages

import at.doml.taskmanager.backend.models.resp.Aliases.{Project, Task, User}
import at.doml.taskmanager.components.{Element, Navbar, Page, Table, TasksTable}
import org.scalajs.dom.document
import scala.scalajs.js
import scala.scalajs.js.JSON
import scala.scalajs.js.annotation.{JSExport, JSExportTopLevel}

@JSExportTopLevel("Projects")
object Projects extends Page with Navbar {

    override val pageButtonId: String = "nav-projects"

    private lazy val usersTable = new Table[User]("project-users") {
        // language=HTML
        override val header: String =
            """
            <th width="250">Name</th>
            <th width="50">Action</th>
            """

        override protected def fillRow(user: User): String = {
            // language=HTML
            s"""
            <td>${user.firstName} ${user.lastName}</td>
            <td>
                <button class="button danger-button small-button" title="Delete user" onclick=""><!-- TODO -->
                    <img height="9" src="img/delete.svg"/>
                </button>
            </td>
            """
        }
    }

    private lazy val inProgressTasksTable = new TasksTable("in-progress-tasks") {
        override protected val firstTimeField: String = "Started"
        override protected val secondTimeField: String = "Due"

        override protected def getFirstTimeField(task: Task): Number = task.startedTimestamp

        override protected def getSecondTimeField(task: Task): Number = task.dueTimestamp

        // language=HTML
        override protected def getButtons: String =
            """
            <button class="button success-button small-button" title="Mark as done">
                <img height="12" src="img/done.svg"/>
            </button>
            <button class="button confirm-button small-button" title="Assign task">
                <img height="12" src="img/take.svg"/>
            </button>
            <button class="button confirm-button small-button" title="Edit task">
                <img height="12" src="img/edit.svg"/>
            </button>
            <button class="button danger-button small-button" title="Delete task" >
                <img height="12" src="img/delete.svg"/>
            </button>
            """
    }

    private lazy val newTasksTable = new TasksTable("new-tasks") {
        override protected val firstTimeField: String = "Created"
        override protected val secondTimeField: String = "Due"

        override protected def getFirstTimeField(task: Task): Number = task.createdTimestamp

        override protected def getSecondTimeField(task: Task): Number = task.dueTimestamp

        // language=HTML
        override protected def getButtons: String =
            """
            <button class="button success-button small-button" title="Start task">
                <img height="12" src="img/start.svg"/>
            </button>
            <button class="button confirm-button small-button" title="Assign task">
                <img height="12" src="img/take.svg"/>
            </button>
            <button class="button confirm-button small-button" title="Edit task">
                <img height="12" src="img/edit.svg"/>
            </button>
            <button class="button danger-button small-button" title="Delete task">
                <img height="12" src="img/delete.svg"/>
            </button>
            """
    }

    private lazy val doneTasksTable = new TasksTable("done-tasks") {
        override protected val firstTimeField: String = "Started"
        override protected val secondTimeField: String = "Finished"

        override protected def getFirstTimeField(task: Task): Number = task.startedTimestamp

        override protected def getSecondTimeField(task: Task): Number = task.finishedTimestamp

        // language=HTML
        override protected def getButtons: String =
            """
            <button class="button confirm-button small-button" title="Re-open task">
                <img height="12" src="img/restart.svg"/>
            </button>
            <button class="button danger-button small-button" title="Delete task">
                <img height="12" src="img/delete.svg"/>
            </button>
            """
    }

    private lazy val usersText = new Element("users-text")
    private lazy val projectTiles = new Element("project-tiles")
    private lazy val projectName = new Element("project-name")
    private lazy val projectDescription = new Element("project-description")
    private lazy val noProject = new Element("no-project")
    private lazy val selectedProject = new Element("selected-project")
    private lazy val inProgressLabel = new Element("in-progress-label")
    private lazy val newLabel = new Element("new-label")
    private lazy val doneLabel = new Element("done-label")
    private lazy val inProgressText = new Element("in-progress-text")
    private lazy val newText = new Element("new-text")
    private lazy val doneText = new Element("done-text")
    private lazy val noTasksLabel = new Element("no-tasks-label")
    private lazy val overviewTab = new Element("overview-tab")
    private lazy val tasksTab = new Element("tasks-tab")
    private lazy val overviewPanel = new Element("overview-panel")
    private lazy val tasksPanel = new Element("tasks-panel")

    private var users: js.Array[User] = js.Array()
    private var projects: js.Array[Project] = js.Array()
    private var currentProject: Project = _
    private var newTasks: js.Array[Task] = js.Array()
    private var doneTasks: js.Array[Task] = js.Array()
    private var inProgressTasks: js.Array[Task] = js.Array()

    override def onNavbarInitSuccess(): Unit = {
        // TODO load from backend
        // language=JSON
        projects = JSON.parse(
            """
            [
                {
                    "id": 1,
                    "name": "project 1",
                    "description": "this is a test project with tasks",
                    "ownerId": 1,
                    "projectStatus": "ACTIVE"
                }, {
                    "id": 2,
                    "name": "project 2",
                    "description": "this is a test project without tasks",
                    "ownerId": 1,
                    "projectStatus": "ACTIVE"
                }
            ]
            """).asInstanceOf[js.Array[Project]]

        projectTiles.empty()

        for (index <- 0 until projects.length) {
            val project = projects(index)

            // language=HTML
            projectTiles.append(
                s"""
                <div id="project-${project.id}" onclick="Projects.selectProject($index)" class="project-tile">
                    <p>${project.name}</p>
                </div>
                """
            )
        }
    }

    @JSExport
    def selectProject(projectIndex: Int): Unit = {
        val project = projects(projectIndex)

        if (currentProject != null && currentProject.id == project.id) {
            return
        }

        if (currentProject != null) {
            document.getElementById(s"project-${currentProject.id}").classList.remove("selected-project-tile")
        }

        currentProject = project

        this.loadTasks()
        this.loadUsers()

        document.getElementById(s"project-${project.id}").classList.add("selected-project-tile")
        projectName.text(project.name)
        projectDescription.text(project.description)
        noProject.noDisplay()
        selectedProject.show()
    }

    private def loadTasks(): Unit = {
        this.clearTasks()

        if (currentProject.id == 1) {
            // language=JSON
            inProgressTasks = JSON.parse(
                """
                [
                    {
                        "id": 555,
                        "priority": 3,
                        "subject": "test task 1",
                        "assigneeId": 1,
                        "createdTimestamp": 1487710619000,
                        "startedTimestamp": 1487710619000,
                        "dueTimestamp": 1487710619000,
                        "finishedTimestamp": 1487710619000,
                        "status": "IN_PROGRESS"
                    }, {
                        "id": 2342,
                        "priority": 2,
                        "subject": "test task 2",
                        "assigneeId": null,
                        "createdTimestamp": 1487710619000,
                        "startedTimestamp": 1487710619000,
                        "dueTimestamp": 1487710619000,
                        "finishedTimestamp": 1487710619000,
                        "status": "IN_PROGRESS"
                    }
                ]
                """).asInstanceOf[js.Array[Task]]
            // language=JSON
            newTasks = JSON.parse(
                """
                [
                    {
                        "id": 235,
                        "priority": 4,
                        "subject": "new task 1",
                        "assigneeId": 1,
                        "createdTimestamp": 1487710619000,
                        "startedTimestamp": 1487710619000,
                        "dueTimestamp": 1487710619000,
                        "finishedTimestamp": 1487710619000,
                        "status": "NEW"
                    }
                ]
                """).asInstanceOf[js.Array[Task]]
            // language=JSON
            doneTasks = JSON.parse(
                """
                [
                    {
                        "id": 999,
                        "priority": 1,
                        "subject": "done task",
                        "assigneeId": 1,
                        "createdTimestamp": 1487710619000,
                        "startedTimestamp": 1487710619000,
                        "dueTimestamp": 1487710619000,
                        "finishedTimestamp": 1487710619000,
                        "status": "DONE"
                    }
                ]
                """).asInstanceOf[js.Array[Task]]
        } else {
            inProgressTasks = new js.Array[Task]()
            newTasks = new js.Array[Task]()
            doneTasks = new js.Array[Task]()
        }
        // TODO load from backend

        if (inProgressTasks.length == 0 && newTasks.length == 0 && doneTasks.length == 0) {
            this.hideTasks()
        } else {
            this.fillTableIfNotEmpty(inProgressTasksTable, inProgressText, inProgressTasks)
            this.fillTableIfNotEmpty(newTasksTable, newText, newTasks)
            this.fillTableIfNotEmpty(doneTasksTable, doneText, doneTasks)
            this.showTasks()
        }
    }

    private def fillTableIfNotEmpty(table: TasksTable, hiddenText: Element, tasks: js.Array[Task]): Unit = {
        if (tasks.length == 0) {
            hiddenText.show()
        } else {
            hiddenText.hide()
            table.fillTable(tasks)
        }
    }

    private def clearTasks(): Unit = {
        inProgressTasksTable.empty()
        newTasksTable.empty()
        doneTasksTable.empty()
    }

    private def hideTasks(): Unit = {
        inProgressTasksTable.hide()
        newTasksTable.hide()
        doneTasksTable.hide()
        inProgressLabel.hide()
        newLabel.hide()
        doneLabel.hide()
        inProgressText.hide()
        newText.hide()
        doneText.hide()
        noTasksLabel.show()
    }

    private def showTasks(): Unit = {
        inProgressTasksTable.show()
        newTasksTable.show()
        doneTasksTable.show()
        inProgressLabel.show()
        newLabel.show()
        doneLabel.show()
        noTasksLabel.hide()
    }

    private def loadUsers(): Unit = {
        usersTable.empty()

        // TODO load from backend
        users = if (currentProject.id == 1) {
            // language=JSON
            JSON.parse(
                """
                [{
                    "id": 11,
                    "username": "user1",
                    "firstName": "User",
                    "lastName": "Test",
                    "role": "ROLE_ADMIN"
                }, {
                    "id": 22,
                    "username": "user2",
                    "firstName": "UUser",
                    "lastName": "TTest",
                    "role": "ROLE_USER"
                }]
                """).asInstanceOf[js.Array[User]]
        } else {
            new js.Array[User]()
        }

        if (users.length == 0) {
            usersTable.hide()
            usersText.show()
        } else {
            usersTable.fillTable(users)
            usersTable.show()
            usersText.hide()
        }
    }

    private def switchTab(newTab: Element, oldTab: Element): Unit = {
        if (!newTab.hasClass("selected-project-tab")) {
            newTab.removeClass("project-tab")
            newTab.addClass("selected-project-tab")
            oldTab.removeClass("selected-project-tab")
            oldTab.addClass("project-tab")
        }
    }

    @JSExport
    def showOverviewTab(): Unit = {
        this.switchTab(overviewTab, tasksTab)

        overviewPanel.show()
        tasksPanel.hide()
    }

    @JSExport
    def showTasksTab(): Unit = {
        this.switchTab(tasksTab, overviewTab)

        tasksPanel.show()
        overviewPanel.hide()
    }
}
